src/misc/degree_sequence.cpp

```diff
@@ -68,66 +70,232 @@ template<typename T> inline bool degree_less(const T &a, const T &b) {
+    // amortized O(alpha(n))
+    igraph_error_t get_spokes(igraph_integer_t degree, const igraph_vector_int_t& seq,
+                              igraph_vector_int_t& spokes) {
+        std::stack<igraph_integer_t> buckets_req;    // stack of needed degree buckets
+        igraph_integer_t num_nodes = 0;
+        igraph_integer_t curr = tail().prev;        // starts with max_bucket
+
+        while (num_nodes < degree && curr > 0) {
+            num_nodes += buckets[curr].count;
+            buckets_req.push(curr);
+            curr = buckets[curr].prev;            // gets next smallest NON-EMPTY bucket
+        }
+        if (num_nodes < degree) {         // not enough spokes for hub degree
+            IGRAPH_ERROR("Degree sequence is not graphical.", IGRAPH_EINVAL);
+        }
+
+        igraph_integer_t num_skip = num_nodes - degree;
+        while (!buckets_req.empty()) {               // starting from smallest degree
+            igraph_integer_t bucket = buckets_req.top();
+            buckets_req.pop();
+
+            igraph_integer_t to_get = buckets[bucket].count - num_skip;
+            while (to_get > 0) {
+                vd_pair node = buckets[bucket].nodes.top();
+                if (VECTOR(seq)[node.vertex] != 0) {                   // if "not marked for removal"
+                    IGRAPH_CHECK(igraph_vector_int_push_back(&spokes, node.vertex)); // add as spoke
+
+                    node.degree--;
+                    insert_node(node);                // first, insert into bucket below
 
-                VECTOR(*edges)[2 * (ec + i)] = vd.vertex;
-                VECTOR(*edges)[2 * (ec + i) + 1] = vertices[i].vertex;
+                    buckets[bucket].count--;
+                    to_get--;
+                }
+                buckets[bucket].nodes.pop();                  // then pop from original bucket
             }
+            num_skip = 0;
         }
+        return IGRAPH_SUCCESS;
+    }
+};
 
-        ec += vd.degree;
+enum class Method {
+    LARGEST_FIRST,
+    SMALLEST_FIRST,
+    CUSTOM
+};
+
+static igraph_error_t igraph_i_havel_hakimi(const igraph_vector_int_t *degseq,
+                                            igraph_vector_int_t *edges,
+                                            Method method,
+                                            const igraph_vector_int_t *hub_order = nullptr) {
+    igraph_integer_t n_nodes = igraph_vector_int_size(degseq);
+
+    // ----- upfront error/graphicality checks ----- //
+    if (method == Method::CUSTOM && hub_order == nullptr) {
+        IGRAPH_ERROR("Custom ordering specified but no order given", IGRAPH_EINVAL);
+    } else if (n_nodes == 0 || (n_nodes == 1 && VECTOR(*degseq)[0] == 0)) {
+        return IGRAPH_SUCCESS;
     }
 
-    return IGRAPH_SUCCESS;
+    for (igraph_integer_t i = 0; i < n_nodes; i++) {
+        igraph_integer_t deg = VECTOR(*degseq)[i];
+        if (deg >= n_nodes) {
+            IGRAPH_ERROR("Invalid degree sequence.", IGRAPH_EINVAL);
```

Comment:
Please use the following message consistently, and make sure that this is done in all _new_ code.

```
"The given degree sequence cannot be realized as a simple graph."
```

Goals:

 - Use the same message whenever the input is not graphical.
 - Communicate which case were're handling (simple graph, multigraph, etc.)